File: cie/evaluators/text_match.py
"""
Text-based evaluator that computes real similarity metrics.
"""

from __future__ import annotations

import json
import math
import statistics
import time
from dataclasses import dataclass
from typing import Any

from cie.config.settings import CIEConfig, get_config
from cie.core.models import Evaluator, Policy, Workload


@dataclass
class TextEvalCase:
    """Single text evaluation case."""

    task: str
    prompt: str
    reference: str
    guidance: str = ""


class TextMatchEvaluator(Evaluator):
    """
    Evaluator that runs small text tasks through the configured model provider and
    scores outputs with Levenshtein distance, similarity ratio, and exact-match heuristics.
    """

    name = "TextMatchEvaluator"

    def __init__(
        self,
        config: CIEConfig | None = None,
        *,
        provider_overrides: dict[str, Any] | None = None,
        cases: list[dict[str, str]] | None = None,
    ):
        self.config = config or get_config()
        self._cases = [TextEvalCase(**case) for case in (cases or self._default_cases())]
        self._provider = self._initialize_provider(provider_overrides or {})

    def run(self, policy: Policy, workload: Workload) -> dict[str, float]:
        """Execute evaluation and compute aggregated metrics."""
        cases = self._resolve_cases(workload)
        latencies_ms: list[float] = []
        similarities: list[float] = []
        exact_matches: list[float] = []
        levenshtein_scores: list[float] = []
        start_time = time.perf_counter()

        for case in cases:
            t0 = time.perf_counter()
            candidate = self._generate_candidate(policy, case)
            latencies_ms.append((time.perf_counter() - t0) * 1000.0)
            metrics = self._compare(candidate, case.reference)
            similarities.append(metrics["similarity"])
            exact_matches.append(metrics["exact_match"])
            levenshtein_scores.append(metrics["levenshtein"])

        elapsed = time.perf_counter() - start_time

        avg_similarity = statistics.fmean(similarities) if similarities else 0.0
        avg_exact = statistics.fmean(exact_matches) if exact_matches else 0.0
        avg_levenshtein = statistics.fmean(levenshtein_scores) if levenshtein_scores else 0.0
        latency_p95 = self._percentile(latencies_ms, 95.0) if latencies_ms else 0.0

        metrics = {
            "latency_p95": latency_p95,
            "cost_per_req": self._estimate_cost(cases, policy),
            "task_success": avg_similarity,
            "context_usage": min(
                1.0, self._average_prompt_tokens(cases) / self.config.model.max_tokens
            ),
            "tool_error_rate": max(0.0, 1.0 - avg_similarity),
            "throughput": len(cases) / elapsed if elapsed else 0.0,
            "memory_usage": 0.35 + 0.05 * len(policy.params),
            "cpu_usage": min(0.95, 0.4 + 0.03 * len(cases)),
            "text_similarity": avg_similarity,
            "text_exact_match": avg_exact,
            "text_levenshtein": avg_levenshtein,
        }
        return metrics

    def get_supported_metrics(self) -> list[str]:
        return [
            "latency_p95",
            "cost_per_req",
            "task_success",
            "context_usage",
            "tool_error_rate",
            "throughput",
            "memory_usage",
            "cpu_usage",
            "text_similarity",
            "text_exact_match",
            "text_levenshtein",
        ]

    def validate_workload(self, workload: Workload) -> bool:
        """Validate workload compatibility."""
        return bool(
            workload.config.get("cases") or workload.config.get("dataset", "text") == "text"
        )

    # Internal helpers --------------------------------------------------------------------- #

    def _initialize_provider(self, overrides: dict[str, Any]) -> Any:
        """Create model provider."""
        try:
            from cie.utils.model_providers import get_model_provider

            model_name = overrides.get("model_name") or self.config.model.model_name
            provider_name = overrides.get("provider") or self.config.model.provider
            api_key = overrides.get("api_key") or self.config.model.api_key
            return get_model_provider(
                provider=provider_name, model_name=model_name, api_key=api_key or "", **overrides
            )
        except Exception:  # pragma: no cover - fallback to mock
            from cie.utils.model_providers import MockModelProvider

            return MockModelProvider(model_name=self.config.model.model_name)

    def _resolve_cases(self, workload: Workload) -> list[TextEvalCase]:
        """Resolve cases from workload config or defaults."""
        config_cases = workload.config.get("cases")
        if config_cases:
            return [TextEvalCase(**case) for case in config_cases]
        return self._cases

    def _generate_candidate(self, policy: Policy, case: TextEvalCase) -> str:
        """Call the provider (or fallback) to generate a candidate string."""
        prompt = self._build_prompt(policy, case)
        try:
            response = self._provider.generate(
                prompt=prompt,
                temperature=self.config.model.temperature,
                max_tokens=min(512, self.config.model.max_tokens),
            )
            if isinstance(response, str):
                return response.strip()
        except Exception:
            pass
        return self._fallback_response(policy, case)

    def _build_prompt(self, policy: Policy, case: TextEvalCase) -> str:
        """Create structured prompt for the evaluator model."""
        params = json.dumps(policy.params, indent=2, sort_keys=True)
        return (
            "You are evaluating an optimization policy. "
            "Produce an answer that best matches the reference text.\n"
            f"Task: {case.task}\n"
            f"Guidance: {case.guidance or 'Follow best practices.'}\n"
            f"Policy parameters:\n{params}\n"
            f"Input:\n{case.prompt}\n"
            "Answer:"
        )

    def _fallback_response(self, policy: Policy, case: TextEvalCase) -> str:
        """Deterministic fallback response if provider is unavailable."""
        seed = f"{policy.name}:{case.task}:{len(policy.params)}"
        truncated = case.reference[: max(10, len(case.reference) // 2)]
        return f"{truncated} :: policy_signature={hash(seed) % 1000}"

    def _compare(self, candidate: str, reference: str) -> dict[str, float]:
        """Compute similarity metrics between two strings."""
        candidate_norm = candidate.strip().lower()
        reference_norm = reference.strip().lower()

        levenshtein = self._levenshtein_distance(candidate_norm, reference_norm)
        max_len = max(len(candidate_norm), len(reference_norm), 1)
        similarity = 1.0 - (levenshtein / max_len)

        exact_match = 1.0 if candidate_norm == reference_norm else 0.0
        return {
            "levenshtein": float(levenshtein),
            "similarity": float(max(0.0, min(1.0, similarity))),
            "exact_match": float(exact_match),
        }

    @staticmethod
    def _levenshtein_distance(a: str, b: str) -> int:
        """Compute Levenshtein distance."""
        if a == b:
            return 0
        if not a:
            return len(b)
        if not b:
            return len(a)
        previous_row = list(range(len(b) + 1))
        for i, ca in enumerate(a, start=1):
            current_row = [i]
            for j, cb in enumerate(b, start=1):
                insertions = previous_row[j] + 1
                deletions = current_row[j - 1] + 1
                substitutions = previous_row[j - 1] + (ca != cb)
                current_row.append(min(insertions, deletions, substitutions))
            previous_row = current_row
        return previous_row[-1]

    @staticmethod
    def _percentile(values: list[float], percentile: float) -> float:
        """Simple percentile implementation."""
        if not values:
            return 0.0
        ordered = sorted(values)
        k = (len(ordered) - 1) * (percentile / 100.0)
        f = math.floor(k)
        c = math.ceil(k)
        if f == c:
            return ordered[int(k)]
        d0 = ordered[int(f)] * (c - k)
        d1 = ordered[int(c)] * (k - f)
        return d0 + d1

    def _average_prompt_tokens(self, cases: list[TextEvalCase]) -> float:
        """Estimate prompt tokens for cases."""
        if not cases:
            return 0.0
        total_chars = sum(len(case.prompt) + len(case.reference) for case in cases)
        return total_chars / max(len(cases), 1)

    def _estimate_cost(self, cases: list[TextEvalCase], policy: Policy) -> float:
        """Rudimentary cost estimation."""
        base_cost = 0.0005 * len(cases)
        params_penalty = 0.0001 * len(policy.params)
        return round(base_cost + params_penalty, 6)

    @staticmethod
    def _default_cases() -> list[dict[str, str]]:
        """Default evaluation cases inspired by Braintrust-style tasks."""
        return [
            {
                "task": "Summarization",
                "prompt": "Summarize the following text in one sentence: "
                "OpenAI released new models focusing on efficiency and multimodality.",
                "reference": "OpenAI announced efficient multimodal models.",
                "guidance": "Be concise and cover the central announcement.",
            },
            {
                "task": "Classification",
                "prompt": "Decide if the customer feedback is positive or negative: "
                "'The UI is clean, but the app crashes every hour.'",
                "reference": "negative",
                "guidance": "Return a single word: positive or negative.",
            },
            {
                "task": "Question Answering",
                "prompt": "Who wrote 'The Pragmatic Programmer'?",
                "reference": "Andy Hunt and Dave Thomas",
                "guidance": "Provide author names separated by 'and'.",
            },
        ]

File: internal/evaluator/textmatch.go
package evaluator

import (
	"fmt"
	"math"
	"strings"
	"time"

	"github.com/cie-team/cie/internal/config"
	"github.com/cie-team/cie/internal/core"
)

// TextMatch scores synthetic “model outputs” with Levenshtein-based similarity.
// No external HTTP calls — outputs are derived from policy parameters for repeatable runs.
type TextMatch struct {
	cfg *config.Config
}

// NewTextMatch constructs the evaluator.
func NewTextMatch(cfg *config.Config) *TextMatch {
	return &TextMatch{cfg: cfg}
}

func (t *TextMatch) Name() string { return "text-match" }

func (t *TextMatch) SupportedMetrics() []string {
	return []string{
		"latency_p95", "cost_per_req", "task_success", "context_usage", "tool_error_rate",
		"throughput", "memory_usage", "cpu_usage",
		"text_similarity", "text_exact_match", "text_levenshtein",
	}
}

func (t *TextMatch) ValidateWorkload(w *core.Workload) bool {
	if w.Config == nil {
		return false
	}
	if _, ok := w.Config["cases"]; ok {
		return true
	}
	ds, _ := w.Config["dataset"].(string)
	return ds == "" || ds == "text"
}

type textCase struct {
	Task      string `json:"task"`
	Prompt    string `json:"prompt"`
	Reference string `json:"reference"`
	Guidance  string `json:"guidance"`
}

func (t *TextMatch) Run(policy *core.Policy, workload *core.Workload) (map[string]float64, error) {
	cases := resolveCases(workload)
	if len(cases) == 0 {
		return nil, fmt.Errorf("text-match: no cases for workload %q", workload.Name)
	}
	start := time.Now()
	var latencies []float64
	var sims, exacts, levs []float64

	quality := policyQuality(policy)

	for _, c := range cases {
		t0 := time.Now()
		cand := synthesizeCandidate(c.Reference, c.Prompt, quality)
		latencies = append(latencies, float64(time.Since(t0).Microseconds())/1000.0)
		cmp := compareStrings(cand, c.Reference)
		sims = append(sims, cmp.similarity)
		exacts = append(exacts, cmp.exact)
		levs = append(levs, cmp.levNorm)
	}

	elapsed := time.Since(start).Seconds()
	avg := mean(sims)
	avgEx := mean(exacts)
	avgLev := mean(levs)
	latP95 := percentile(latencies, 95)

	maxTok := 4096.0
	if t.cfg != nil && t.cfg.Model.MaxTokens > 0 {
		maxTok = float64(t.cfg.Model.MaxTokens)
	}
	tokEst := averagePromptLen(cases)

	return map[string]float64{
		"latency_p95":      latP95,
		"cost_per_req":     0.0005 + 0.00002*float64(len(cases)),
		"task_success":     avg,
		"context_usage":    min64(1, tokEst/maxTok),
		"tool_error_rate":  max64(0, 1-avg),
		"throughput":       float64(len(cases)) / math.Max(elapsed, 1e-6),
		"memory_usage":     0.35 + 0.05*float64(len(policy.Params)),
		"cpu_usage":        min64(0.95, 0.4+0.03*float64(len(cases))),
		"text_similarity":  avg,
		"text_exact_match": avgEx,
		"text_levenshtein": avgLev,
	}, nil
}

func resolveCases(w *core.Workload) []textCase {
	if w.Config == nil {
		return defaultTextCases()
	}
	raw, ok := w.Config["cases"]
	if !ok {
		if ds, _ := w.Config["dataset"].(string); ds == "text" || ds == "" {
			return defaultTextCases()
		}
		return nil
	}
	arr, ok := raw.([]any)
	if !ok {
		return nil
	}
	var out []textCase
	for _, it := range arr {
		m, ok := it.(map[string]any)
		if !ok {
			continue
		}
		tc := textCase{
			Task:      asString(m["task"]),
			Prompt:    asString(m["prompt"]),
			Reference: asString(m["reference"]),
			Guidance:  asString(m["guidance"]),
		}
		if tc.Reference != "" {
			out = append(out, tc)
		}
	}
	if len(out) == 0 {
		return defaultTextCases()
	}
	return out
}

func defaultTextCases() []textCase {
	return []textCase{
		{Task: "Sentiment", Prompt: "positive or negative?", Reference: "positive"},
		{Task: "Short answer", Prompt: "95th latency metric?", Reference: "latency_p95"},
	}
}

func asString(v any) string {
	switch x := v.(type) {
	case string:
		return x
	case fmt.Stringer:
		return x.String()
	default:
		return ""
	}
}

func policyQuality(p *core.Policy) float64 {
	if p.Params == nil {
		return 0.55
	}
	q := 0.55
	if v, ok := floatParam(p.Params, "temperature"); ok {
		q += (1.0 - v) * 0.25
	}
	if v, ok := floatParam(p.Params, "prune_ratio"); ok {
		q -= v * 0.15
	}
	if v, ok := floatParam(p.Params, "learning_rate"); ok {
		q += min64(0.15, v)
	}
	return math.Max(0.05, math.Min(0.98, q))
}

func synthesizeCandidate(reference, prompt string, quality float64) string {
	// Corrupt reference proportionally to (1-quality)
	r := []rune(reference)
	if len(r) == 0 {
		return prompt[:min(len(prompt), 12)]
	}
	steps := int(math.Ceil(float64(len(r)) * (1 - quality) * 0.8))
	out := append([]rune(nil), r...)
	for i := 0; i < steps && len(out) > 0; i++ {
		op := (i + len(prompt)) % 3
		switch op {
		case 0: // delete
			out = append(out[:i%len(out)], out[i%len(out)+1:]...)
		case 1: // substitute
			pos := i % len(out)
			out[pos] = rune('a' + (i % 26))
		default: // insert
			pos := i % max(1, len(out))
			ch := rune(' ')
			if len(prompt) > 0 {
				ch = rune(prompt[i%len(prompt)])
			}
			out = append(out[:pos], append([]rune{ch}, out[pos:]...)...)
		}
		if len(out) == 0 {
			break
		}
	}
	return string(out)
}

type cmp struct {
	similarity float64
	exact      float64
	levNorm    float64
}

func compareStrings(a, b string) cmp {
	if a == b {
		return cmp{similarity: 1, exact: 1, levNorm: 1}
	}
	d := levenshtein([]rune(strings.ToLower(a)), []rune(strings.ToLower(b)))
	maxLen := max(len([]rune(a)), len([]rune(b)))
	if maxLen == 0 {
		return cmp{}
	}
	levNorm := 1.0 - float64(d)/float64(maxLen)
	levNorm = math.Max(0, math.Min(1, levNorm))
	// similarity: Jaccard on word tokens + lev blend
	sim := 0.6*levNorm + 0.4*jaccardWords(a, b)
	return cmp{similarity: sim, exact: 0, levNorm: levNorm}
}

func jaccardWords(a, b string) float64 {
	ta := wordSet(a)
	tb := wordSet(b)
	if len(ta) == 0 && len(tb) == 0 {
		return 1
	}
	inter := 0
	for w := range ta {
		if tb[w] {
			inter++
		}
	}
	union := len(ta) + len(tb) - inter
	if union == 0 {
		return 0
	}
	return float64(inter) / float64(union)
}

func wordSet(s string) map[string]bool {
	m := map[string]bool{}
	for _, w := range strings.Fields(strings.ToLower(s)) {
		if w != "" {
			m[w] = true
		}
	}
	return m
}

func levenshtein(a, b []rune) int {
	if len(a) == 0 {
		return len(b)
	}
	if len(b) == 0 {
		return len(a)
	}
	prev := make([]int, len(b)+1)
	for j := 0; j <= len(b); j++ {
		prev[j] = j
	}
	for i := 1; i <= len(a); i++ {
		cur := make([]int, len(b)+1)
		cur[0] = i
		for j := 1; j <= len(b); j++ {
			cost := 0
			if a[i-1] != b[j-1] {
				cost = 1
			}
			cur[j] = minInt(
				cur[j-1]+1,
				minInt(prev[j]+1, prev[j-1]+cost),
			)
		}
		prev = cur
	}
	return prev[len(b)]
}

func minInt(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func mean(xs []float64) float64 {
	if len(xs) == 0 {
		return 0
	}
	s := 0.0
	for _, x := range xs {
		s += x
	}
	return s / float64(len(xs))
}

func percentile(xs []float64, p float64) float64 {
	if len(xs) == 0 {
		return 0
	}
	// naive copy sort
	cp := append([]float64(nil), xs...)
	for i := 0; i < len(cp); i++ {
		for j := i + 1; j < len(cp); j++ {
			if cp[j] < cp[i] {
				cp[i], cp[j] = cp[j], cp[i]
			}
		}
	}
	idx := int(math.Ceil(float64(len(cp))*p/100)) - 1
	if idx < 0 {
		idx = 0
	}
	if idx >= len(cp) {
		idx = len(cp) - 1
	}
	return cp[idx]
}

func averagePromptLen(cases []textCase) float64 {
	if len(cases) == 0 {
		return 0
	}
	s := 0.0
	for _, c := range cases {
		s += float64(len(c.Prompt) + len(c.Reference))
	}
	return s / float64(len(cases))
}

func max64(a, b float64) float64 {
	if a > b {
		return a
	}
	return b
}

func min64(a, b float64) float64 {
	if a < b {
		return a
	}
	return b
}
